### DISCUSSION TITLE
Future of AIVE

### DISCUSSION THREAD

**Discussion by Mogli12** [Mogli12/AIVehicleExtension#272] [Ideas]
Hi everybody,

I am thinking about rebuilding AIVE in FS22. The current version is just a conversion of the FS19 version. But that version did have some drawbacks.

And the name does not fit anymore. The new specialization is AIFieldWorker. Thus, I am looking for a new name. How about AdvancedFieldWorker (AFW)?

I am thinking of scanning the fields at the beginning of the job. The mod could calculate lanes and headland lanes. There are differen strategies how to do the headland:
* Start on circular headland (combine and sowing machines)
* Start of first lane, continue lane by lane and do the headland lane by lane at the end (plowing). Difficulty is the space outside of the field for the turn maneuver at the end of headland lanes.
* Skip left and right lanes at beginning and work on every N-th lane until finished. Do circular headland at the end
* Special mode for beet plows?
* Drive to discharge of refill point at field border?

I want to do this with a minimum set of configuration steps. In best case to drive to one corner of the field, press the right button and everything works automatically.

Best regards

**Comment by Dovynight** [comment]
Hey, Advanced Field Worker sounds great tbh. Can't wait until i can try this mod out on fs22. Your mod was a life saver in fs17 and 19. Vanilla AI just sucks.. I hope you won't abondon your mod man, good luck.

**Comment by rasc1974** [comment]
Hi, honestly I use AiVE more or less only for plowing in straight lines but with headland. And this works great! For all the other stuff (e.g. spraying, combining, cultivating) Courseplay actually is for me the preferred option. However, if the new tool will provide an interface with a very simple handling it will also be interesting! I´m excited!

**Comment by Mogli12** [comment]
If I recreate AIVE, then it should be different to Courseplay.

**Comment by omid900** [comment]
Personally i use AIVE for work in pair with the ai vehicles, for ex. ploughing a field using 2 tractors, one behind and offset the other. also when combining, so whatever you do with as rework version of AIVE, i hope it remains with a feature of using an offset working width setting.  About the name, i think advance field worker is a good one ;)

**Comment by Mogli12** [comment]
Most of the source code can be found in the FS22 installation folder `./sdk/debugger/gameSource.zip`. I checked how the spec. AIFieldWorker and AIFieldJob belong together.

This is my conclusion: Alternative field jobs are started via the ingame menu only. Probably, there is a input action to start the alternative field job with default settings. But I think that `KEY_h`starts the standard field worker.

I think that the start of a field job generates a course around the field. Parts of the course that are perpendicular to the current direction are headland. The other parts are right and left border of the inner up-/down-rows. The corners of the field need special attention. I think it could be solved like in this video: https://youtu.be/gRBRFSed4RI?t=119. A similar example for harvesters is here: https://youtu.be/bpmeAguBTEk?t=410.

The principle is always the same. If the tool is in the back, then you drive until end of field without leaving the field. You turn by 180°, drive again to the end of the field and finish the 1st track. Afterwards, you start the next row. If the tool is in the front, then you do it in opposite way. You harvest until end of track, drive reverse and harvest the start of the next, opposite track. Afterwards, you turn by 180° and continue on the next track.

As alternative, the left and right border rows together with the headland rows could be done in circular mode. It really depends on the tools. Harvesters and sowing machines would do the border rows and headland in circular mode. Gray mowing would to the headland first and then start with the middle row to allow wild animals to leave the field. Plowing happens row by row.

**Comment by Mogli12** [comment]
If you want to work with multiple vehicles at the same time, e.g. two combine harvesters, then all vehicles should be treated as one combination. There are special requirements regarding turning radius. The 1st vehicle will always do the left most row in driving direction.

**Comment by Kynuska** [comment]
Hello.
And he would go to the mod to control the player, not just the assistant. As in real agriculture in modern tractors with GPS. I load the field boundaries, determine the strategy and control the vehicle along the lines.
A functional ninimap (as in CaC) would then be a nice icing on the cake and replace the Control Panel.
It's only a matter of time before the SIM Dashboard screen appears in the future.
For example, a randomly searched video:
https://www.youtube.com/watch?v=SpUHYrV5Cq0&ab_channel=JohnDeere

Best regards
Kynuska

**Comment by Mogli12** [comment]
I think SIM dashboard works via keyboard key combinations, right? That sounds difficult...

**Reply by Kynuska** [reply]
I didn't realize he probably wouldn't be able to get the map out of the game. But it does not matter.

**Comment by regs1980** [comment]
Mit neuem Elan und Motivation durch die gute Bewertung im Modhub verwirkliche mal den AdvancedFieldWorker, kann ein interessantes Projekt werden, wenn man das so ließt was du dir so vorstellst.

**Comment by Williwillswisse** [comment]
Great idea. Start the field job by sending the vehicle to the field via AD and then handover to AFW to do the field work. After work switch over to AD again and drive to the park position of the vehicle.

**Comment by Mogli12** [comment]
I made some progress. I am able to create my own field worker job in the in-game menu. The job itself is not yet interesting. It is more a less a simple copy of the standard (straight) field worker. 

The first new feature is to restart the field worker in single player after loading a save game. And I am able to add parameters to the right column of the in-game menu.
![fsScreen_2022_01_20_14_19_10](https://user-images.githubusercontent.com/8790374/150346419-2fedc7fc-3ca4-474d-b64f-2075dede832c.png)

My ideas about interoperability are:
- Export AFW job as courseplay course?
- Define an additional point&angle to fill or discharge the vehicle. The vehicle would automatically drive to this point if empty or full.

Regarding using AD to send the vehicle to a field. The standard field worker already includes a go-to task at the beginning. And I not so sure if it easy to replace it by AD.

There is at least one ask about multiple vehicles (combines) on one field. The difficulty about this is that jobs are saved at the vehicle. Maybe it could work to save the job at the 1st vehicle. All additional vehicles will just save their state plus the link the to 1st vehicle.

**Comment by regs1980** [comment]
How is the project progressing?
